lib/schemes/oauth2.js

```diff
@@ -198,4 +202,85 @@ export default class Oauth2Scheme {
 
     return true // True means a redirect happened
   }
+
+  // ---------------------------------------------------------------
+  // Watch axios requests for token expiration
+  // Refresh tokens if token has expired
+  // ---------------------------------------------------------------
+
+  initializeTokenRefreshOnRequest () {
+    const { $axios } = this.$auth.ctx.app
+    let isRefreshing = false
+
+    $axios.onRequest(async config => {
+      let token = this.$auth.getToken(this.name)
+      let refreshToken = this.$auth.getRefreshToken(this.name)
+
+      // Token or "refresh token" does not exists
+      if (!token || !refreshToken || !token.length || !refreshToken.length) {
+        return config
+      }
+
+      // If already trying to refresh token, do not try again
+      if (isRefreshing) {
+        return config
+      }
+
+      // Time variables
+      let tokenExpiresAt = jwtDecode(token).exp * 1000
+      let refreshTokenExpiresAt = jwtDecode(refreshToken).exp * 1000
+      const now = Date.now()
+
+      // Give us some slack to help the token from expiring between validation and usage
+      const timeSlackMillis = 500
+      tokenExpiresAt -= timeSlackMillis
+      refreshTokenExpiresAt -= timeSlackMillis
+
+
+      // Return if token has not expired
+      if (now < tokenExpiresAt) {
+        return config
+      }
+      // "Refresh token" has also expired. There is no way to refresh. Force logout.
+      if (now > refreshTokenExpiresAt) {
+        this.logout()
+
+        if(process.client) {
+          // Explicitly redirect to the signed-out page.
+          // We don't want to redirect a user with an expired token on page reload/first navigation.
+          // WatchLoggedIn in auth.js->mounted redirects to logout when the state changes,
+          // but it only works in the client
+          this.$auth.redirect('logout')
+        }
+        // Stop the request from happening. The original caller must catch ExpiredSessionErrors
+        throw new ExpiredSessionError()
+      }
+
+      // Try to refresh token before processing current request
+      isRefreshing = true
+
+      return $axios.post(this.options.access_token_endpoint,
+        encodeQuery({
+          refresh_token: refreshToken.replace(this.options.token_type + ' ', ''),
+          client_id: this.options.client_id,
+          grant_type: 'refresh_token'
+        })
+      ).then(response => {
+        isRefreshing = false
+
+        // Update token and "refresh token"
+        token = this.options.token_type + ' ' + response.data.access_token
```

Comment:
Maybe we can refactor (`token_type` prefix) to util inside the auth class and also use it for schemes?

Reply:
Good idea! Something like this?
```js
// auth.js
addTokenPrefix(token) {
  return this.options.token_type + ' ' + token
}
```
Then we would have to add token_type as a cross-scheme default. It's defined in both local.js and oauth2.js, so options.token_type is not defined in auth.js. What do you think, @pi0 ? 🙂